prodsys/simulation/path_finder.py

```diff
@@ -0,0 +1,218 @@
+from typing import List, TYPE_CHECKING, Optional, Union
+from pathfinding.core.graph import Graph, GraphNode
+from pathfinding.finder.dijkstra import DijkstraFinder
+
+from prodsys.simulation import request
+
+if TYPE_CHECKING:
+    from prodsys.simulation import resources, request, sink, source, process
+
+class Pathfinder:
+    """
+    Class representing a path finder for transportation requests in a graph.
+    """
+
+    def __init__(self):
+        """
+        Initializes two empty lists of nodes and node locations.
+        """
+        self.nodes: List[GraphNode] = []
+        self.node_loc: List[Union[resources.NodeData, sink.Sink, source.Source]] = []
+
+    def find_path(self, request: request.TransportResquest, which_path: bool, self_obj):
+        """
+        The general function which includes all sub functions to find the shortest path for a TransportRequest.
+
+        Args:
+            request (TransportResquest): The transportation request.
+            which_path (bool): Indicates whether to find the path from AGV to origin or from origin to target.
+            self_obj: The process of the related resource
+
+        Returns:
+            The path as a list of links.
+        """
+        given_links_list = self_obj.links
+
+        edges = self.process_links_to_edges(request, given_links_list=given_links_list)
+        graph = Graph(edges=edges, bi_directional=True)
+        origin, target = self.origin_target_to_graphnode(request, graph, which_path, given_links_list=given_links_list)
+        if origin is None or target is None:
+            return False
+        g_path = self.find_graphnode_path(origin, target, graph)
+        path = self.node_path_to_link_path(g_path, request, given_links_list=given_links_list)
+
+        return path
+
+    def process_links_to_edges(self, request: request.TransportResquest, given_links_list):
+        """
+        Processes the given links to create (Graph)-edges for the graph.
+
+        Args:
+            request (TransportResquest): The transportation request.
+            given_links_list: The given links list.
+
+        Returns:
+            The edges as a list of tuples (with a start_node, end_node and related costs).
+        """
+        from prodsys.simulation import resources, sink, source
+        pathfinder_edges = []
+
+        for link in given_links_list:
```

Comment:
Do less nesting here in the code -> move logical bits to separate functions, to make code more readable

Reply:
Commit 26